rlpx.md

```diff
@@ -42,118 +60,305 @@ the authenticity of the message by checking whether
 
 Initiator:
-egress-mac = keccak256.update(mac-secret ^ recipient-nonce || auth-sent-init)
-# destroy nonce
-ingress-mac = keccak256.update(mac-secret ^ initiator-nonce || auth-recvd-ack)
-# destroy remote-nonce
+
+```
+egress-mac = keccak256.init((mac-secret ^ recipient-nonce) || auth)
+ingress-mac = keccak256.init((mac-secret ^ initiator-nonce) || ack)
+```
 
 Recipient:
-egress-mac = keccak256.update(mac-secret ^ initiator-nonce || auth-sent-ack)
-# destroy nonce
-ingress-mac = keccak256.update(mac-secret ^ recipient-nonce || auth-recvd-init)
-# destroy remote-nonce
+
+```
+egress-mac = keccak256.init((mac-secret ^ initiator-nonce) || ack)
+ingress-mac = keccak256.init((mac-secret ^ recipient-nonce) || auth)
 ```
 
-Creating authenticated connection:
+When a frame is sent, the corresponding MAC values are computed by updating the
+`egress-mac` state with the data to be sent. The update is performed by XORing the header
+with the encrypted output of it's corresponding MAC. This is done to ensure uniform
+operations are performed for both plaintext MAC and ciphertext. All MACs are sent
+cleartext.
 
-1. initiator connects to recipient and sends `auth` message
-2. recipient accepts, decrypts and verifies `auth` (checks that recovery of signature ==
-   `keccak256(ephemeral-pubk)`)
-3.  recipient generates `auth-ack` message from `remote-ephemeral-pubk` and `nonce`
-4.  recipient derives secrets and sends the first payload frame
-5.  initiator receives `auth-ack` and derives secrets
-6.  initiator sends first payload frame
-7.  recipient receives and authenticates first payload frame
-8.  initiator receives and authenticates first payload frame
-9.  cryptographic handshake is complete if MAC of first payload frame is valid on both sides
+```text
+header-mac-seed = aes(mac-secret, keccak256.digest(egress-mac)[:16] ^ header-ciphertext)
+egress-mac = keccak256.update(egress-mac, header-mac-seed)
+header-mac = keccak256.digest(egress-mac)[:16]
+```
+
+Computing `frame-mac`:
+
+```text
+egress-mac = keccak256.update(egress-mac, frame-ciphertext)
+frame-mac-seed = aes(mac-secret, keccak256.digest(egress-mac)[:16]) ^ keccak256.digest(egress-mac)[:16]
+egress-mac = keccak256.update(egress-mac, frame-mac-seed)
+frame-mac = keccak256.digest(egress-mac)[:16]
+```
+
+Verifying the MAC on ingress frames is done by updating the `ingress-mac` state in the
+same way as `egress-mac` and comparing to the values of `header-mac` and `frame-mac` in
+the ingress frame. This should be done before decrypting `header-ciphertext` and
+`frame-ciphertext`.
+
+# Capability Messaging
 
-# Framing
+All messages following the initial handshake are associated with a 'capability'. Any
+number of capabilities can be used concurrently on a single RLPx connection.
 
-All packets following `auth` are framed. Either side may disconnect if authentication of
-the first framed packet fails.
+A capability is identified by an ASCII string and version number. The capabilities
```

Comment:
I would suggest adding a couple of examples, eg: eth, les, Istanbul and note that while there is no length limit, the shorter the better.